File: examples/OozieWorkflowSpark/src/main/java/org/apache/spark/examples/WordCount.java
/**
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */


package org.apache.spark.examples;

import java.util.ArrayList;
import java.util.List;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;

import scala.Tuple2;

public class WordCount {
    public static void main(String[] args) {
        if (args.length < 2) {
            System.err.println("Usage: WordCountSparkMain <file> <file>");
            System.exit(1);
        }
        String inputPath = args[0];
        String outputPath = args[1];
        SparkConf sparkConf = new SparkConf().setAppName("Word count");
        try (JavaSparkContext ctx = new JavaSparkContext(sparkConf)) {
            JavaRDD<String> lines = ctx.textFile(inputPath, 1);
            JavaRDD<String> words = lines.flatMap(new FlatMapFunction<String, String>() {
                private static final long serialVersionUID = 1L;
                public Iterable<String> call(String sentence) {
                    List<String> result = new ArrayList<>();
                    if (sentence != null) {
                        String[] words = sentence.split(" ");
                        for (String word : words) {
                            if (word != null && word.trim().length() > 0) {
                                result.add(word.trim().toLowerCase());
                            }
                        }
                    }
                    return result;
                }
            });
            JavaPairRDD<String, Integer> pairs = words.mapToPair(new PairFunction<String, String, Integer>() {
                private static final long serialVersionUID = 1L;
                public Tuple2<String, Integer> call(String s) {
                    return new Tuple2<>(s, 1);
                }
            });
 
            JavaPairRDD<String, Integer> counts = pairs.reduceByKey(new Function2<Integer, Integer, Integer>() {
                private static final long serialVersionUID = 1L;
                public Integer call(Integer a, Integer b) {
                    return a + b;
                }
            }, 2);
            JavaPairRDD<Integer, String> countsAfterSwap = counts.mapToPair(new PairFunction<Tuple2<String, Integer>, Integer, String>() {
                private static final long serialVersionUID = 2267107270683328434L;
                @Override
                public Tuple2<Integer, String> call(Tuple2<String, Integer> t)
                        throws Exception {
                    return new Tuple2<>(t._2, t._1);
                }
            });
            countsAfterSwap = countsAfterSwap.sortByKey(false);
            counts = countsAfterSwap.mapToPair(new PairFunction<Tuple2<Integer, String>, String, Integer>() {
                private static final long serialVersionUID = 2267107270683328434L;
                @Override
                public Tuple2<String, Integer> call(Tuple2<Integer, String> t)
                        throws Exception {
                    return new Tuple2<>(t._2, t._1);
                }
            });
            JavaRDD<String> results = counts.map(new Function<Tuple2<String, Integer>, String>() {
                @Override
                public String call(Tuple2<String, Integer> v1) throws Exception {
                    return String.format("%s,%s", v1._1, Integer.toString(v1._2));
                }
            });
            results.saveAsTextFile(outputPath);
        }
    }
}

File: examples/CloudantIntegrationWithSpark/wordcount.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from __future__ import print_function

import time
import sys
from operator import add

from pyspark import SparkContext
from pyspark.sql import SQLContext

# create a Cloudant database using httplib
# httplib is used because other libraries aren't available on BigInsights by default
# This code will be made redundant by: https://github.com/cloudant-labs/spark-cloudant/pull/59/commits/01234e09cf71faca82b8dcaf67cb423c3206ad8c
def createDatabase(hostname, username, password, dbname):
    import httplib
    import base64
    import string

    auth = base64.encodestring('%s:%s' % (username, password)).replace('\n', '')
    restservice = httplib.HTTPS(hostname)
    restservice.putrequest("PUT", '/' + dbname)
    restservice.putheader("Host", hostname)
    restservice.putheader("Authorization", "Basic %s" % auth)
    restservice.endheaders()
    restservice.send('')
    statuscode, statusmessage, header = restservice.getreply()

    if statuscode != 201:
        raise Exception(statuscode, stausmessage)

if __name__ == "__main__":
    if len(sys.argv) != 6:
        print("Usage: wordcount <file> <cloudant hostname> <cloudant username> <cloudant password> <cloudant database>", file=sys.stderr)
        exit(-1)

    license_filename = sys.argv[1]
    cl_hostname = sys.argv[2]
    cl_username = sys.argv[3]
    cl_password = sys.argv[4]
    cl_database = sys.argv[5]

    createDatabase(cl_hostname, cl_username, cl_password, cl_database)
   
    sc = SparkContext(appName="PythonWordCount")

    # read file from HDFS
    lines = sc.textFile(license_filename, 1)
    counts = lines.flatMap(lambda x: x.split(' ')) \
                  .map(lambda x: (x, 1)) \
                  .reduceByKey(add) \
                  .filter(lambda x: x[0].isalnum())

    sqlContext = SQLContext(sc)

    # Create a sql dataframe from the counts dataframe    
    df = sqlContext.createDataFrame(counts,['letter', 'count'])

    # Let's peek at the schema and some records
    print('*' * 80)
    print(df.printSchema())
    print(df.show(20))
    print('*' * 80)
    
    # Write the data to Cloudant
    df.write.format("com.cloudant.spark") \
        .option("cloudant.host",cl_hostname) \
        .option("cloudant.username",cl_username) \
        .option("cloudant.password",cl_password) \
        .save(cl_database)

    print("Word counts of LICENSE file have been saved to https://{0}/{1}".format(cl_hostname, cl_database))

    sc.stop()
